src/libtorch.cc

```diff
@@ -1155,19 +1192,81 @@ ModelInstanceState::ProcessRequests(
-              request_count, &responses, &compute_end_ns));
+      if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+#ifdef TRITON_ENABLE_GPU
+        at::cuda::CUDAStream stream =
+            at::cuda::getStreamFromExternal(stream_, DeviceId());
+        at::cuda::CUDAStreamGuard device_guard{stream};
+        RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+            responses, request_count, all_response_failed,
+            ReadOutputTensors(
+                total_batch_size, output_tensors, requests, request_count,
+                &responses));
+#endif
+      } else {
+        RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+            responses, request_count, all_response_failed,
+            ReadOutputTensors(
+                total_batch_size, output_tensors, requests, request_count,
+                &responses));
+      }
     }
   }
 
   uint64_t exec_end_ns = 0;
-  SET_TIMESTAMP(exec_end_ns);
+  if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+#ifdef TRITON_ENABLE_GPU
+    RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+        responses, request_count, all_response_failed,
+        ConvertCUDAStatusToTritonError(
+            cudaEventRecord(compute_output_end_event_, stream_),
+            TRITONSERVER_ERROR_INTERNAL, "Failed to record the event"));
+    RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+        responses, request_count, all_response_failed,
+        ConvertCUDAStatusToTritonError(
+            cudaEventSynchronize(compute_output_end_event_),
+            TRITONSERVER_ERROR_INTERNAL, "Failed to synchronize event"));
+    float compute_input_duration = 0;
+    float compute_infer_duration = 0;
+    float compute_output_duration = 0;
+    RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+        responses, request_count, all_response_failed,
+        ConvertCUDAStatusToTritonError(
+            cudaEventElapsedTime(
+                &compute_input_duration, compute_input_event_,
+                compute_infer_event_),
+            TRITONSERVER_ERROR_INTERNAL, "Failed to capture elapsed time"));
+
+    RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+        responses, request_count, all_response_failed,
+        ConvertCUDAStatusToTritonError(
+            cudaEventElapsedTime(
+                &compute_infer_duration, compute_infer_event_,
+                compute_output_event_),
+            TRITONSERVER_ERROR_INTERNAL, "Failed to capture elapsed time"));
+
+    RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+        responses, request_count, all_response_failed,
+        ConvertCUDAStatusToTritonError(
+            cudaEventElapsedTime(
+                &compute_output_duration, compute_output_event_,
+                compute_output_end_event_),
+            TRITONSERVER_ERROR_INTERNAL, "Failed to capture elapsed time"));
+    compute_start_ns = exec_start_ns + (compute_input_duration * 1e6);
+    compute_end_ns = compute_start_ns + (compute_infer_duration * 1e6);
+    exec_end_ns = compute_end_ns + (compute_output_duration * 1e6);
```

Comment:
I think `exec_end_ns` can just be the timestamp captured at this place and no CUDA event is needed. Then we can move the calculation after the responses are sent

Reply:
I agree. I have fixed this.